### PR TITLE
feat(liquidation-reserve-currency): add reserve currency liquidator to bots and refine smart contract

### LINKED ISSUE #2751: Ability for bots to only need to hold a single reserve currency versus  holding collateral and tokens for each EMP
Created by Chris Maree via monday.com integration. 🎉

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

issue #2751

**Summary**

This PR integrates the `SwapMintLiquidate` contract added in PR #2750 into the liquidator bot. Additionally, this PR modifies the original contract to consider any existing synthetics or collateral that the DSProxy might already have.

**Details**

The base of the implementation is built around a new module added to the bots: a `ProxyTransactionWrapper`. this wrapper abstracts the action of sending liquidation transactions away from the implementation thereby separating concerns between liquidation decision making and transaction sending. 

This module implements two main functions:

1. `getEffectiveSyntheticTokenBalance` which returns the balance of the DSProxy in synthetics + the total amount of synthetics the bot _could_ have, considering the collateral and reserve currencies based off a swap & mint transaction. This is used when computing WDF logic. If the bot is running in normal EOA mode then this simply returns the current synthetic balance.
2. `submitLiquidationTransaction` which sends a liquidation tx. This is either a normal EOA liquidation (as before) or is routed via the DSProxy liquidator.

The code change in this PR is pretty large. I've tried to identify key areas of interest and draw attention for the reviewer to places where the most complex/important logic is located.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [X]  All existing tests pass
- [ ]  Untested

**Issue(s)**

close #2751

### CODE REVIEW

File: packages/affiliates/payouts/2021-03-15_2021-03-22_0019.json [hunk 1]
```diff
@@ -6,187 +6,45 @@
     "startTime": 1615849200000,
     "endTime": 1616454000000,
     "empWhitelist": [
-      [
```
Comment (chrismaree) [inline packages/affiliates/payouts/2021-03-15_2021-03-22_0019.json:9]: unrelated the main PR. changes came due to running `yarn lint-fix`

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 2]
```diff
@@ -2,6 +2,7 @@ pragma solidity ^0.6.0;
 pragma experimental ABIEncoderV2;
```
Comment (chrismaree) [inline packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:2]: This contract was basically re-written with a bunch more logic within it. In particular, the reserve currency liquidator now considers any existing synthetic or colalteral tokens the DSproxy might have. For example, if we manually sent the DSProxy a bunch of synthetic tokens it should use these before minting extras.

File: packages/core/contracts/bot-helpers/uniswap-broker/UniswapBroker.sol [hunk 3]
```diff
@@ -24,7 +24,7 @@ contract UniswapBroker {
      * @dev True price is expressed in the ratio of token A to token B.
      * @dev The caller must approve this contract to spend whichever token is intended to be swapped.
      * @param tradingAsEOA bool to indicate if the UniswapBroker is being called by a DSProxy or an EOA.
-     * @param uniswapRouter address of the uniswap router used to facilate trades.
```
Comment (chrismaree) [inline packages/core/contracts/bot-helpers/uniswap-broker/UniswapBroker.sol:27]: fixed some typos.

File: packages/core/test/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.js [hunk 4]
```diff
@@ -64,6 +64,19 @@ contract("ReserveTokenLiquidator", function(accounts) {
   const sponsor2 = accounts[2];
```
Comment (chrismaree) [inline packages/core/test/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.js:64]: These tests were heavily extended to accommodate the new functionality.

File: packages/disputer/test/Disputer.js [hunk 5]
```diff
@@ -73,7 +73,7 @@ const _setFundingRateAndAdvanceTime = async fundingRate => {
 
 // If the current version being executed is part of the `supportedVersions` array then return `it` to run the test.
 // Else, do nothing. Can be used exactly in place of a normal `it` to parameterize contract types and versions supported
-// for a given test.eg: versionedversionedIt([{ contractType: "any", contractVersion: "any" }])(["Perpetual-latest"])("test name", async function () { assert.isTrue(true) })
```
Comment (chrismaree) [inline packages/disputer/test/Disputer.js:76]: there was this common typo throughout the repo. unrelated to main PR.

File: packages/financial-templates-lib/src/proxy-transaction-handler/DSProxyManager.js [hunk 6]
```diff
@@ -24,11 +24,6 @@ class DSProxyManager {
     this.dsProxyAddress = null;
 
     // Helper functions from web3.
```
Comment (chrismaree) [inline packages/financial-templates-lib/src/proxy-transaction-handler/DSProxyManager.js:26]: unrelated to main code change. this utility had unused declerations.

File: packages/liquidator/index.js [hunk 7]
```diff
@@ -196,9 +204,40 @@ async function run({
     const gasEstimator = new GasEstimator(logger);
     await gasEstimator.update();
 
+    const dsProxyManager = new DSProxyManager({
```
Comment (chrismaree) [inline packages/liquidator/index.js:207]: main change here is that we now need to instantiate the `dsProxyManager` and the `proxyTransaction` wrapper to enable. The latter is fed into the liquidator as a parameter.

File: packages/liquidator/index.js [hunk 8]
```diff
@@ -337,7 +376,22 @@ async function Poll(callback) {
       startingBlock: process.env.STARTING_BLOCK_NUMBER,
       // Block number to search for events to. If set, acts to limit from where the monitor bot will search for events up
       // until. If either startingBlock or endingBlock is not sent, then the bot will search for event.
-      endingBlock: process.env.ENDING_BLOCK_NUMBER
+      endingBlock: process.env.ENDING_BLOCK_NUMBER,
+      // If enabled, the bot will funnel liquidations via a DSProxy which will be deployed on the bots behalf. This
```
Comment (chrismaree) [inline packages/liquidator/index.js:380]: how the DSProxy is used is configurable through envs, as per usual.

File: packages/liquidator/src/liquidator.js [hunk 9]
```diff
@@ -205,7 +205,9 @@ class Liquidator {
   async liquidatePositions(maxTokensToLiquidateWei, liquidatorOverridePrice) {
     this.logger.debug({
       at: "Liquidator",
-      message: "Checking for liquidatable positions and preforming liquidations"
+      message: "Checking for liquidatable positions and preforming liquidations",
```
Comment (chrismaree) [inline packages/liquidator/src/liquidator.js:208]: improved this log a bit by adding the params send in.

File: packages/liquidator/index.js [hunk 10]
```diff
@@ -250,7 +289,7 @@ async function run({
           if (!isExpiredOrShutdown) {
             // Check for liquidatable positions and submit liquidations. Bounded by current synthetic balance and
             // considers override price if the user has specified one.
-            const currentSyntheticBalance = await syntheticToken.methods.balanceOf(accounts[0]).call();
```
Comment (chrismaree) [inline packages/liquidator/index.js:253]: note this change: previously the first parram fed into the `liquidatePositions` method was the current synthetic token balance of the liquidator. Now, it uses the `proxyTransactionWrapper`'s `getEffectiveSyntheticTokenBalance` method. This method considers: a) how many tokens the DSProxy already has b) how much collateral the DSProx has, and how many synths could be minted from this collateral and c) how much collateral the DSProxy has, how much collateral this could buy, and how many synthetics this could mint. i.e an "effective" balance that is available to the bot at the moment of liquidation.

File: packages/liquidator/src/liquidator.js [hunk 11]
```diff
@@ -269,23 +271,22 @@ class Liquidator {
...
       const liquidationArgs = this.liquidationStrategy.processPosition({
-        // position is assumed to be undercollateralized
-        position,
-        // need to know our current balance to know how much to save for defense fund
-        syntheticTokenBalance,
-        currentBlockTime,
-        // this is required to create liquidation object
-        maxCollateralPerToken,
-        // maximum tokens we can liquidate in position
+        position, // position is assumed to be undercollateralized
```
Comment (chrismaree) [inline packages/liquidator/src/liquidator.js:284]: styling nit. no need to use new lines here.

File: packages/liquidator/src/liquidator.js [hunk 12]
```diff
@@ -310,79 +311,31 @@ class Liquidator {
...
+      this.logger.debug({
+        at: "Liquidator",
+        message: "Liquidating position",
+        position,
+        inputPrice: scaledPrice.toString(),
+        minLiquidationPrice: this.liquidationMinPrice,
+        maxLiquidationPrice: maxCollateralPerToken.toString(),
+        tokensToLiquidate: tokensToLiquidate.toString()
+      });
+      const logResult = await this.proxyTransactionWrapper.submitLiquidationTransaction(liquidationArgs);
```
Comment (chrismaree) [inline packages/liquidator/src/liquidator.js:323]: this is the primary change to this file. previously, the `liquidator` would submit the liquidation inline here. see the code removed 👇 .

After this PR, the liquidator simply calls the `.proxyTransactionWrapper.submitLiquidationTransaction(liquidationArgs)` which handles all logic associated with the actual liquidation transaction.

File: packages/liquidator/src/liquidator.js [hunk 13]
```diff
@@ -310,79 +311,31 @@ class Liquidator {
...
-        // Receipt without events
-        receipt = await ynatm.send({
-          sendTransactionFunction: gasPrice => liquidation.send({ ...txnConfig, nonce, gasPrice }),
-          minGasPrice,
-          maxGasPrice,
-          gasPriceScalingFunction,
-          delay: 60000 // Tries and doubles gasPrice every minute if tx hasn't gone through
-        });
-      } catch (error) {
+      if (logResult instanceof Error || !logResult)
```
Comment (chrismaree) [inline packages/liquidator/src/liquidator.js:325]: log depending on the output.

File: packages/merkle-distributor/scripts/example.json [hunk 14]
```diff
@@ -37,4 +37,4 @@
       "metaData": { "reason": ["Governance Votes KPI Options Round 1"] }
     }
   }
-}
```
Comment (chrismaree) [inline packages/merkle-distributor/scripts/example.json:40]: change due to lint.

File: yarn.lock [hunk 15]
```diff
@@ -1363,7 +1363,7 @@
   dependencies:
```
Comment (chrismaree) [inline yarn.lock:1363]: not 100% sure why the diff here is so large. This simply came about after running a yarn install. Should I revert this file changes @UMAprotocol/eng ?

File: packages/liquidator/test/Liquidator.js [hunk 16]
```diff
@@ -1365,8 +1395,7 @@ contract("Liquidator.js", function(accounts) {
               );
               await financialContract.setCurrentTime(nextTime);
 
-              // Liquidator only has enough balance to liquidate 1 position fully, will
-              // minimally liquidate the other
+              // Liquidator only has enough balance to liquidate 1 position fully, will minimally liquidate the other
```
Comment (chrismaree) [inline packages/liquidator/test/Liquidator.js:1398]: nit: line length can be wrapped.

File: packages/liquidator/test/Liquidator.js [hunk 17]
```diff
@@ -1659,6 +1686,464 @@ contract("Liquidator.js", function(accounts) {
             }
           );
         });
+        describe("Liquidation via DSProxy", () => {
```
Comment (chrismaree) [inline packages/liquidator/test/Liquidator.js:1689]: pretty in depth tests. please let me know if there is an edge case I dont cover!

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+        { rawValue: this.maxReserverTokenSpent }, // maxReserverTokenSpent
+        { rawValue: liquidationArgs[1].rawValue }, // minCollateralPerTokenLiquidated
+        { rawValue: liquidationArgs[2].rawValue }, // maxCollateralPerTokenLiquidated. This number need to be >= the token price.
+        { rawValue: liquidationArgs[3].rawValue }, // maxTokensToLiquidate. This is how many tokens the positions has (liquidated debt).
+        liquidationArgs[4]
+      )
+      .encodeABI();
+    const callCode = this.ReserveCurrencyLiquidator.bytecode;
+
+    const dsProxyCallReturn = await this.dsProxyManager.callFunctionOnNewlyDeployedLibrary(callCode, callData);
```
Comment (chrismaree) [inline packages/liquidator/src/proxyTransactionWrapper.js:261]: note that this does not use `ynatm`. a separate PR should be added to implement this gas escalation within the `dsProxyManager`. I added a tasks to track this and will work on it as soon as this is done with code review: https://github.com/UMAprotocol/protocol/issues/2788

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+        this.web3.utils.isAddress(this.liquidatorReserveCurrencyAddress),
+        "Must provide a reserve currency address to use the proxy transaction wrapper!"
+      );
+    }
+
+    this.reserveToken = new web3.eth.Contract(getAbi("ExpandedERC20"), this.liquidatorReserveCurrencyAddress);
+    this.ReserveCurrencyLiquidator = getTruffleContract("ReserveCurrencyLiquidator", web3, "latest");
+  }
+
+  // TODO: wrap this into a common util.
```
Comment (chrismaree) [inline packages/liquidator/src/proxyTransactionWrapper.js:110]: This needs to be refactored. I did not put it in this PR as it is in a few other places. I added an issue to track this task here: https://github.com/UMAprotocol/protocol/issues/2789

File: packages/liquidator/src/liquidator.js [hunk 20]
```diff
@@ -393,13 +346,29 @@ class Liquidator {
       message: "Checking for expired and disputed liquidations to withdraw rewards from"
     });
 
+    // In legacy versions of the EMP, withdrawing needs to be done by a party involved in the liquidation (i.e liquidator,
+    // sponsor or disputer).As the liquidator is the DSProxy, we would require the ability to send the withdrawal tx
+    // directly from the DSProxy to facilitate this.This functionality is not implemented as almost all legacy EMPs expired.
+    if (this.proxyTransactionWrapper.isUsingDsProxyToLiquidate && this.isLegacyEmpVersion) {
```
Comment (chrismaree) [inline packages/liquidator/src/liquidator.js:352]: @UMAprotocol/eng reviewers: not this decision here. If the bot is using DSProxy and the contract is legacy then withdrawing rewards is disabled. I chose to do this to keep the overall code churn and feature addition as simple as possible, especially seeing that almost all legacy contracts have expired. If a liquidation happens on a legacy contract we can still liquidate. However, we will need to manually withdraw from the DSProxy.

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 21]
```diff
@@ -39,44 +42,64 @@ contract ReserveCurrencyLiquidator {
...
+        // will factor this in. If the DSProxy has more collateral than the total amount required the purchased = 0.
+        FixedPoint.Unsigned memory collateralToBePurchased =
+            subOrZero(
+                totalCollateralRequired,
+                FixedPoint.Unsigned(IERC20(fc.collateralCurrency()).balanceOf(address(this)))
+            );
+
+        // 4.a. If there is some collateral to be purchased, execute a trade on uniswap to meet the shortfall.
+        // Note the path assumes a direct route from the reserve currency to the collateral currency.
+        if (collateralToBePurchased.isGreaterThan(FixedPoint.fromUnscaledUint(0))) {
```
Comment (nicholaspai) [inline packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:79]: What if `fc.collateralCurrency() == reserveCurrency`?

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 22]
```diff
@@ -39,44 +42,64 @@ contract ReserveCurrencyLiquidator {
         address financialContract,
         address reserveCurrency,
         address liquidatedSponsor,
-        FixedPoint.Unsigned calldata maxReserverTokenSpent,
+        FixedPoint.Unsigned calldata maxReserveTokenSpent,
         FixedPoint.Unsigned calldata minCollateralPerTokenLiquidated,
         FixedPoint.Unsigned calldata maxCollateralPerTokenLiquidated,
         FixedPoint.Unsigned calldata maxTokensToLiquidate,
         uint256 deadline
     ) public {
```
Comment (nicholaspai) [inline packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:50]: I thought about reentrancy vulnerabilities here and I don't think there's any issue since this contract doesn't store any internal state, but just want to make sure you've thought about that too and perhaps you add a reentrancy guard if its a concern.

File: packages/liquidator/index.js [hunk 23]
```diff
@@ -49,7 +51,8 @@ async function run({
   liquidatorConfig,
   liquidatorOverridePrice,
   startingBlock,
-  endingBlock
+  endingBlock,
+  proxyTransactionWrapperConfig
```
Comment (nicholaspai) [inline packages/liquidator/index.js:55]: Can you add comments about this param?

File: packages/liquidator/index.js [hunk 24]
```diff
@@ -196,9 +202,37 @@ async function run({
...
+    let dsProxyManager;
+    if (proxyTransactionWrapperConfig?.isUsingDsProxyToLiquidate) {
+      dsProxyManager = new DSProxyManager({
+        logger,
+        web3,
+        gasEstimator,
+        account: accounts[0],
+        dsProxyFactoryAddress:
+          proxyTransactionWrapperConfig?.dsProxyFactoryAddress || getAddress("DSProxyFactory", networkId),
+        dsProxyFactoryAbi: getAbi("DSProxyFactory"),
```
Comment (nicholaspai) [inline packages/liquidator/index.js:214]: Since these two ABI's are hardcoded, I think it might be cleaner to call `getAbi()` from within the DXProxyManager. I do this with the `funding-rate-proposer/src/proposer.js` client, but I'm not strongly opionated.

File: packages/liquidator/index.js [hunk 25]
```diff
@@ -196,9 +202,37 @@ async function run({
...
+        dsProxyAbi: getAbi("DSProxy")
+      });
+
+      await dsProxyManager.initializeDSProxy();
```
Comment (nicholaspai) [inline packages/liquidator/index.js:218]: What does this function do? Might be worth adding a comment about which side effects trigger

File: packages/liquidator/src/liquidator.js [hunk 26]
```diff
@@ -269,23 +271,22 @@ class Liquidator {
 
       // Note: query the time again during each iteration to ensure the deadline is set reasonably.
       const currentBlockTime = this.financialContractClient.getLastUpdateTime();
-      const syntheticTokenBalance = this.toBN(await this.syntheticToken.methods.balanceOf(this.account).call());
+
+      // TODO: right now the `processPosition` method takes in maxTokensToLiquidateWei and syntheticTokenBalance. These
```
Comment (nicholaspai) [inline packages/liquidator/src/liquidator.js:275]: Don't you think it could be useful to distinguish between the bot's current balance and the amount of balance its allocating to liquidate with? I.e. it allows the caller to set a limit on how many tokens to liquidate

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 27]
```diff
@@ -0,0 +1,288 @@
...
+    let truffleContractCreator = truffleContract(contractJsonObject);
+    truffleContractCreator.setProvider(web3.currentProvider);
+    return truffleContractCreator;
+  }
+
+  // Get the effective synthetic token balance. If the bot is executing in normal mode (liquidations sent from an EOA)
+  // then this is simply the token balance of the unlocked account. If the liquidator is using a DSProxy to liquidate,
+  // then consider the synthetics could be minted, + any synthetics the DSProxy already has.
+  async getEffectiveSyntheticTokenBalance() {
+    const syntheticTokenBallance = await this.syntheticToken.methods.balanceOf(this.account).call();
```
Comment (nicholaspai) [inline packages/liquidator/src/proxyTransactionWrapper.js:121]: nit: mispelled `balance`

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 28]
```diff
@@ -0,0 +1,288 @@
...
+   * */
+  constructor({
+    web3,
+    financialContract,
+    gasEstimator,
+    syntheticToken,
+    collateralToken,
+    account,
+    dsProxyManager = undefined,
+    isUsingDsProxyToLiquidate = false,
```
Comment (nicholaspai) [inline packages/liquidator/src/proxyTransactionWrapper.js:38]: Maybe i'm trying to over-refactor, but why can't we remove the `isUsingDsProxyToLiquidate` config boolean, and just check if `dsProxyManager == undefined`? It seems below on lines 94, 122,  and 182 that you are checking the former but I feel like you can also just check for `!dsProxymanager`

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 29]
```diff
@@ -0,0 +1,288 @@
...
+
+  // Main entry point for submitting a liquidation. If the bot is not using a DSProxy then simply send a normal EOA tx.
+  // If the bot is using a DSProxy then route the tx via it.
+  async submitLiquidationTransaction(liquidationArgs) {
+    // If the liquidator is not using a DSProxy, use the old method of liquidating
+    if (!this.isUsingDsProxyToLiquidate) return await this._executeLiquidationWithoutDsProxy(liquidationArgs);
+    else return await this._executeLiquidationWithDsProxy(liquidationArgs);
+  }
+
+  async _executeLiquidationWithoutDsProxy(liquidationArgs) {
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/liquidator/src/proxyTransactionWrapper.js:186]: Maybe this isn't the PR to do this but we should probably refactor the `ynatm` retry logic into `common/src/ContractUtils.js:runTransaction()`, which currently manages transactions for the `funding-rate-proposer` and `optimistic-oracle-proposer`. This would remove the `estimateGas-send` code below.

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 18 above)
Comment (nicholaspai) [reply packages/liquidator/src/proxyTransactionWrapper.js:261]: Can we refactor some of this into a generalized function in  `common/src/ContractUtils`?

**Review by nicholaspai — COMMENTED** [review]
Tons of good code good job, left some comments for my first pass. Once you address I'll go through again.

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 18 above)
Comment (chrismaree) [reply packages/liquidator/src/proxyTransactionWrapper.js:261]: yup! I will change the `dsProxyManager` to use this. 

However, it looks like your use of the `runTransaction` in the proposer does not use ynatm either. we should ideally modify the run transaction to use this so we consistently bump gas all over the place.

File: packages/liquidator/src/liquidator.js (hunk 20 above)
Comment (chrismaree) [reply packages/liquidator/src/liquidator.js:352]: @nicholaspai I think it's fine as there are basically no contracts running on the legacy version.

@mrice32 yup! we would need a very simple script to do this. I added a simple smart contract to this PR that we could use, if the need arises. it implements the correct interface for this contract version:

```
pragma solidity ^0.6.0;
pragma experimental ABIEncoderV2;

import "../../common/implementation/FixedPoint.sol";

// Simple contract used to withdraw liquidations using a DSProxy from legacy contracts (1.2.2 and below).
contract LiquidationWithdrawer {
    function withdrawLiquidation(
        address financialContractAddress,
        uint256 liquidationId,
        address sponsor
    ) public returns (IFinancialContract.RewardsData memory) {
        return IFinancialContract(financialContractAddress).withdrawLiquidation(liquidationId, sponsor);
    }
}

interface IFinancialContract {
    function withdrawLiquidation(uint256 liquidationId, address sponsor)
        external
        returns (FixedPoint.Unsigned memory amountWithdrawn);
}
```

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol [hunk 30]
```diff
@@ -39,44 +42,64 @@ contract ReserveCurrencyLiquidator {
...
+            subOrZero(maxTokensToLiquidate, FixedPoint.Unsigned(IERC20(fc.tokenCurrency()).balanceOf(address(this))));
 
-        // 4. Mint synthetics with collateral. Note we are minting at the GCR and minting the exact number liquidated.
-        TransferHelper.safeApprove(fc.collateralCurrency(), address(fc), totalCollateralNeeded.rawValue);
-        fc.create(collateralToMintAtGcr, maxTokensToLiquidate);
-
-        // 5. Liquidate position with newly minted synthetics.
+        // 2. Calculate how much collateral is needed to make up the token shortfall from minting new synthetics.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
+        FixedPoint.Unsigned memory collateralToMintShortfall = tokenShortfall.mul(gcr);
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:60]: this is a great point @nicholaspai! I dont think there is any risk to change this to divCeil other than there might be a bit of dust left over in the contract. this is a strictly better state than the contract having a shortfall of synthetics.

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol (hunk 21 above)
Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:79]: This is a great question. I'm going to write some unit tests to verify this but my understanding is that the trade will just not happen. might be better to have some branching logic

Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:79]: I added a test which validates this. Simply no trade happens and everything else continues on as it was before. no additional branching is needed; the `swapTokenForExactTokens` simply does nothing and returns :)

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol (hunk 22 above)
Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:50]: ye, not too worried about it. this contract is called from a DSProxy and so the context of state is the DSProxy storage. This contract should never hold funds either.

File: packages/liquidator/index.js (hunk 24 above)
Comment (chrismaree) [reply packages/liquidator/index.js:214]: Ideally I dont want to modify the `DSProxyManager` in this PR though as I feel that'll be covering too much. I'll do it in a follow up PR where I refactor the `DSProxyManager` to use the `runTransaction`.

File: packages/liquidator/index.js (hunk 25 above)
Comment (chrismaree) [reply packages/liquidator/index.js:218]: it will load in an existing DSProxy for the account EOA if one already exists or create a new one for the user. I added a comment

File: packages/liquidator/src/liquidator.js (hunk 26 above)
Comment (chrismaree) [reply packages/liquidator/src/liquidator.js:275]: ye, this is something I spent quite a lot of time thinking about. Previously, this parameter was simply the bots token balance. Within the `liquidatePositions` method it also read in the current token balance. Then, the WDF logic took in two parameters that were _the same_.

What I did here is implement this to be the _same_ as the previous implementation with the goal of not modifying key logic that we know to work within the WDF. In a separate PR we should refactor this. There were a few other things in how the WDF is called and the data it returns that I'd also like to change.

I felt it best to mimic the current logic as much as posible and leave these changes for later. do you agree?

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 28 above)
Comment (chrismaree) [reply packages/liquidator/src/proxyTransactionWrapper.js:38]: you could do that. the reason I did it like this is that I want the user to explicitly define a `useDsProxyToLiquidate=true` within the `DSPROXY_CONFIG` and by doing it like this I can simply pass down the `proxyTransactionWrapperConfig` into the wrapper. if I modified it the way you described I'd need to destruct the config within the index and do some logic on it rather than just passing it around>

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:79]: good point. done!

File: packages/liquidator/src/proxyTransactionWrapper.js [hunk 31]
```diff
@@ -0,0 +1,288 @@
...
+  createContractObjectFromJson(contractJsonObject) {
+    let truffleContractCreator = truffleContract(contractJsonObject);
+    truffleContractCreator.setProvider(web3.currentProvider);
+    return truffleContractCreator;
+  }
+
+  // Get the effective synthetic token balance. If the bot is executing in normal mode (liquidations sent from an EOA)
+  // then this is simply the token balance of the unlocked account. If the liquidator is using a DSProxy to liquidate,
+  // then consider the synthetics could be minted, + any synthetics the DSProxy already has.
+  async getEffectiveSyntheticTokenBalance() {
```
[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/liquidator/src/proxyTransactionWrapper.js:120]: great point and a good refactor. it would be posible to return say the "effective" balance from the smart contract and then call this as a view function. However, this contract is already at the limit of size and any extra returns cause it to bow over the limit :(

For now I think its ok to leave like this, if you ok with that?

File: packages/liquidator/src/liquidator.js (hunk 20 above)
Comment (nicholaspai) [reply packages/liquidator/src/liquidator.js:352]: Is this tested?

File: packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol (hunk 30 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol:60]: Agreed

File: packages/liquidator/src/liquidator.js (hunk 26 above)
Comment (nicholaspai) [reply packages/liquidator/src/liquidator.js:275]: Yes agreed, minimize churn for this PR.

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 29 above)
[the remark this answers is not in the document]
Comment (nicholaspai) [reply packages/liquidator/src/proxyTransactionWrapper.js:186]: I'll work on this in a spearate PR

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 18 above)
Comment (nicholaspai) [reply packages/liquidator/src/proxyTransactionWrapper.js:261]: I'm going to work on this in another PR

**Review by nicholaspai — COMMENTED** [review]
Thanks for reviewing my comments, my only large comment remaining is to add tests to the LiquidationWithdrawer.sol contract

File: packages/liquidator/src/proxyTransactionWrapper.js (hunk 31 above)
[the remark this answers is not in the document]
Comment (mrice32) [reply packages/liquidator/src/proxyTransactionWrapper.js:120]: Yeah, that's cool with me. Are you saying the library smart contract is at the bytecode limit? That's kinda hard to believe....

[the remark this answers is not in the document]
Comment (chrismaree) [reply packages/liquidator/src/proxyTransactionWrapper.js:120]: it hits a stack too deep error with any additional state vars or any return types.

**Review by mrice32 — APPROVED** [review]
